Reorganizing DGMulti mesh types

jlchan opened the pull request:
Todo: 
- [x] Replace `VertexMappedMesh` with `DGMultiMesh` 
- [x] Introduce `CartesianMesh` constructor
- [x] Make sure periodic SBP operators with a single-element mesh work correctly (their domain can only be set in the mesh constructor)

This PR is intended to clean up `DGMulti` meshes. Currently, everything is stored in a `VertexMappedMesh` mesh type because I didn't know what I was doing when I introduced it :sweat_smile:. The name `VertexMappedMesh` refers to "elements whose mapping is determined only by vertex positions", but the name is misleading. The type itself is just a small wrapper around `MeshData`, which stores geometric change of variables information, and a list of domain boundaries.

To clean this up, I've instead renamed `VertexMappedMesh` -> `DGMultiMesh`, and am introducing constructors for each instance. You can still call`VertexMappedMesh` with the same arguments, but it will just return an instance of `DGMultiMesh` instead now. 

I plan on adding additional type parameters to specialize on different types of meshes in the next PR.

ranocha commented:
The docs should be updated, too - for example, "Simplicial mesh" should probably become "DGMulti mesh"?

jlchan commented:
> The docs should be updated, too - for example, "Simplicial mesh" should probably become "DGMulti mesh"?

Yes, definitely. Thanks for the reminder.

sloede commented on docs/src/meshes/dgmulti_mesh.md:
```diff
@@ -98,21 +98,21 @@ the `examples/dgmulti_2d` and `examples/dgmulti_3d` folders. Some key elixirs to
 contains a `Dict` of boundary segment names (symbols) and list of faces which lie on that boundary
 segment.
 
+A `DGMultiMesh` can be constructed in several ways. For example, `CartesianMesh(dg::DGMulti)` will
+return a Cartesian mesh on ``[-1, 1]^d`` with element types specified by `dg`.
+`DGMulti` meshes can also be constructed by specifying a list
+of vertex coordinates `vertex_coordinates_x`, `vertex_coordinates_y`, `vertex_coordinates_z` and a
+connectivity matrix `EToV` where `EToV[e,:]` gives the vertices which correspond to element `e`.
+These quantities are available from most unstructured mesh generators.
+
```
```suggestion
A `DGMultiMesh` can be constructed in several ways. For example, `CartesianMesh(dg::DGMulti)` will
return a Cartesian `DGMultiMesh` on ``[-1, 1]^d`` with element types specified by `dg`.
`DGMultiMesh`es can also be constructed by specifying a list
of vertex coordinates `vertex_coordinates_x`, `vertex_coordinates_y`, `vertex_coordinates_z` and a
connectivity matrix `EToV` where `EToV[e,:]` gives the vertices which correspond to element `e`.
These quantities are available from most unstructured mesh generators.

```
Maybe a little clearer this way that all DGMuli Meshes are of type `DGMultiMesh`?

sloede commented on NEWS.md:
```diff
@@ -16,6 +16,10 @@ for human readability.
+#### Removed
+
+- The `VertexMappedMesh` type was removed in favor of the `DGMultiMesh` type.
+  The `VertexMappedMesh` constructor is still supported but will be deprecated.
```
Why "will be" and not "is"?

sloede commented on NEWS.md:
```diff
@@ -16,6 +16,10 @@ for human readability.
 
+#### Removed
+
+- The `VertexMappedMesh` type was removed in favor of the `DGMultiMesh` type.
```
```suggestion
- The `VertexMappedMesh` type is removed in favor of the `DGMultiMesh` type.
```
Should it be present tense, as in "these are the changes that are applied with the new release" (as opposed to the more passive "these changes have been applied during this release cycle")? I always flip flop on this though, so just make sure you stick to the prevailing style ;-)

jlchan replied:
I have no idea TBH 😅. I'll go with present tense if it's consistent with previous announcements.

sloede commented on src/Trixi.jl:
```diff
@@ -204,7 +204,8 @@ export trixi_include, examples_dir, get_examples, default_example, default_examp
 
-export DGMulti, AbstractMeshData, VertexMappedMesh, estimate_dt, CartesianMesh
+export DGMulti, DGMultiMesh, estimate_dt, CartesianMesh
+export VertexMappedMesh # TODO: deprecate
```
Actually deprecate it and remove todo?

sloede commented on src/meshes/dgmulti_meshes.jl:
```diff
@@ -54,41 +53,44 @@ end
 end
 
+# TODO: deprecate this interface
+VertexMappedMesh(args...; kwargs...) = DGMultiMesh(args...; kwargs...)
```
Why not deprecated?

sloede commented on src/solvers/dgmulti/types.jl:
```diff
@@ -76,25 +76,50 @@ function DGMulti(element_type::AbstractElemShape,
+  DGMultiMesh(triangulateIO, dg.basis, boundary_dict)
+
+"""
+  CartesianMesh(dg::DGMulti; cells_per_dimension,
```
What's the motivation for introducing a new name `CartesianMesh` instead of making it just another `DGMultiMesh` constructor? IIRC, we do something similar for the `StructuredMesh` or `P4estMesh`, which can also be constructed as a Cartesian mesh. If I remember correctly, `CartesianMesh` is only used for a convenience ctor and not for dispatch, so there doesn't seem to be "hard" motivation for a new type. I might be mistaken though...

ranocha replied:
Sounds reasonable to use the same name (`DGMultiMesh`) if there won't be another underlying type :+1:

jlchan replied:
I introduced `CartesianMesh` mainly to match https://github.com/trixi-framework/Trixi.jl/pull/1010. I'm fine with swapping it for a `DGMultiMesh` constructor. 

I could use some feedback though - I will need to dispatch on the mesh type eventually. I can do this by adding type parameters to `DGMultiMesh` (e.g., `DGMultiMesh{NDIMS, ..., Cartesian}` or `DGMultiMesh{NDIMS, ..., Curved}`) or just by introducing new mesh types and making `DGMultiMesh` an abstract type (e.g., `CartesianMesh <: AbstractDGMultiMesh`). Any preference?

ranocha replied:
I don't have a strong opinion on this

sloede replied:
There's to little for me to go on, to make a confident suggestion. For this PR, I'd stick with not creating a separate type.

For the future, I would probably use an explicit type rather than just a type parameter. This allows you to choose different (more efficient/cleaner?) implementations for simpler mesh types (such as Cartesian) and makes it easier to read (and write) for dispatch. But I'd leave this discussion for the PR in which you actually have to introduce them ;-)

jlchan replied:
OK, this sounds good. I'll save the introduction of a new type for later, but I think I should then rename some of the variables to emphasize what assumptions existing routines are making based on `DGMulti` mesh types. This should just be a renaming 

1. `AbstractMeshData` -> `AbstractDGMultiMesh` and `DGMultiMesh` -> `AffineMesh` 
2. add `AffineMesh` type annotations whenever I assume constant geometric terms.

I'll then introduce `CurvedMesh <: AbstractDGMultiMesh` in a future PR.

sloede reviewed (changes requested):
Overall it LGTM, but I've left a few questions

ranocha commented on docs/src/meshes/dgmulti_mesh.md:
```diff
@@ -98,21 +98,21 @@ the `examples/dgmulti_2d` and `examples/dgmulti_3d` folders. Some key elixirs to
 contains a `Dict` of boundary segment names (symbols) and list of faces which lie on that boundary
 segment.
 
+A `DGMultiMesh` can be constructed in several ways. For example, `CartesianMesh(dg::DGMulti)` will
```
```suggestion
A [`DGMultiMesh`](@ref) can be constructed in several ways. For example, `CartesianMesh(dg::DGMulti)` will
```

ranocha commented on examples/dgmulti_1d/elixir_euler_flux_diff.jl:
```diff
@@ -11,9 +11,7 @@ equations = CompressibleEulerEquations1D(1.4)
-cells_per_dimension = (8, )
-vertex_coordinates, EToV = StartUpDG.uniform_mesh(dg.basis.elementType, cells_per_dimension...)
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg, is_periodic=(true,))
+mesh = CartesianMesh(dg, cells_per_dimension=(8,), is_periodic=(true,))
```
While we're already talking about name changes and deprecations: Would it make sense to unify the argument names indicating periodic meshes? Currently, we use `periodic` for our Kopriva-style DGSEM meshes and `is_periodic` for our Hesthaven-Warburton-style `DGMulti` meshes.

sloede replied:
I am all for consistency. IIRC, we had a discussion about the naming here before, but I don't remember any details. I think back then it was between "periodic" and "periodicity". I don't have a strong opinion here, but would tend to favor the status quo (="periodic") if just for avoiding additional work.

ranocha replied:
IIRC, ishould be `periodicity` to math the other mesh types - please check and deprecate the old keyword.

ranocha replied:
Did you also deprecate the old keyword? You can do this as in
https://github.com/ranocha/SummationByPartsOperators.jl/blob/659ccd22ddd54964c70ad3648d384f3f2324384f/src/periodic_operators.jl#L637-L639 
for example

jlchan replied:
@ranocha I don't understand how `Base.depwarn` would work in this case. In https://github.com/ranocha/SummationByPartsOperators.jl/blob/659ccd22ddd54964c70ad3648d384f3f2324384f/src/periodic_operators.jl#L637-L639, you seem to have different argument types for `mode`, but here the argument type stays the same, only the name changes.

jlchan replied:
Nvm, this seems to work https://discourse.julialang.org/t/deprecate-keyword-argument/33740.

jlchan replied:
OK, I think I got it working in https://github.com/trixi-framework/Trixi.jl/pull/1012/commits/e690f7d49c359bc0f8c234ed89f77e1e47d7d2c7 now.

jlchan commented:
@ranocha @sloede I think I've addressed most of the comments. Once this and https://github.com/trixi-framework/Trixi.jl/pull/1014 are merged, I'll submit one more PR to introduce some new `DGMulti` mesh types before the next release.

ranocha commented on examples/dgmulti_2d/elixir_euler_brown_minion_vortex.jl:
```diff
@@ -28,10 +28,9 @@ function initial_condition_BM_vortex(x, t, equations::CompressibleEulerEquations
-vertex_coordinates, EToV = StartUpDG.uniform_mesh(dg.basis.elementType, cells_per_dimension...)
-vertex_coordinates = map(x -> 0.5 .* x, vertex_coordinates) # map domain to [-0.5, 0.5]^2
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg, is_periodic=(true, true))
+mesh = DGMultiMesh(dg, cells_per_dimension=(16, 16),
+                     coordinates_min=(-0.5, -0.5), coordinates_max=(0.5, 0.5),
+                     periodicity=(true, true))
```
```suggestion
mesh = DGMultiMesh(dg, cells_per_dimension=(16, 16),
                   coordinates_min=(-0.5, -0.5), coordinates_max=(0.5, 0.5),
                   periodicity=(true, true))
```

ranocha commented on examples/dgmulti_2d/elixir_euler_rayleigh_taylor_instability.jl:
```diff
@@ -65,15 +65,9 @@ dg = DGMulti(polydeg = 3, element_type = Quad(), approximation_type = Polynomial
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg; is_periodic=(true,false))
+mesh = DGMultiMesh(dg, cells_per_dimension=(num_elements, 4 * num_elements),
+                     coordinates_min=(0.0, 0.0), coordinates_max=(0.25, 1.0),
+                     periodicity=(true,false))
```
```suggestion
                     periodicity=(true, false))
```

ranocha commented on examples/dgmulti_2d/elixir_euler_weakform_periodic.jl:
```diff
@@ -9,9 +9,7 @@ equations = CompressibleEulerEquations2D(1.4)
-cells_per_dimension = (4, 4)
-vertex_coordinates, EToV = StartUpDG.uniform_mesh(dg.basis.elementType, cells_per_dimension...)
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg, is_periodic=(true,true))
+mesh = DGMultiMesh(dg, cells_per_dimension = (4, 4), periodicity=(true,true))
```
```suggestion
mesh = DGMultiMesh(dg, cells_per_dimension = (4, 4), periodicity=(true, true))
```

ranocha commented on examples/dgmulti_2d/elixir_mhd_weak_blast_wave.jl:
```diff
@@ -15,11 +15,9 @@ dg = DGMulti(polydeg=3, element_type = Quad(), approximation_type = Polynomial()
-vertex_coordinates = map(x -> 2 * x, vertex_coordinates) # map domain to [-2, 2]^2
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg, is_periodic=(true, true))
-
+mesh = DGMultiMesh(dg, cells_per_dimension=(16, 16),
+                     coordinates_min=(-2, -2), coordinates_max=(2, 2),
+                     periodicity=(true, true))
```
```suggestion
mesh = DGMultiMesh(dg, cells_per_dimension=(16, 16),
                   coordinates_min=(-2.0, -2.0), coordinates_max=(2.0, 2.0),
                   periodicity=(true, true))
```
This clarifies that we're using `Float64` here.

ranocha commented on examples/dgmulti_2d/elixir_mhd_weak_blast_wave_SBP.jl:
```diff
@@ -19,10 +19,9 @@ dg = DGMulti(polydeg=3, element_type = Quad(), approximation_type = SBP(),
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg, is_periodic=(true, true))
-
+mesh = DGMultiMesh(dg, cells_per_dimension=cells_per_dimension,
+                     coordinates_min=(-2, -2), coordinates_max=(2, 2),
```
```suggestion
                     coordinates_min=(-2.0, -2.0), coordinates_max=(2.0, 2.0),
```

ranocha commented on examples/dgmulti_2d/elixir_shallowwater_source_terms.jl:
```diff
@@ -14,10 +14,10 @@ dg = DGMulti(polydeg=3, element_type = Quad(), approximation_type = SBP(),
-vertex_coordinates = map(x -> sqrt(2) .* 0.5 .* (1 .+ x), vertex_coordinates) # map domain to [0, sqrt(2)]^2
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg, is_periodic=(true, true))
+mesh = DGMultiMesh(dg, cells_per_dimension=(8, 8),
+                     coordinates_min=(0, 0), coordinates_max=(sqrt(2), sqrt(2)),
```
```suggestion
                     coordinates_min=(0.0, 0.0), coordinates_max=(sqrt(2), sqrt(2)),
```

ranocha commented on examples/dgmulti_3d/elixir_euler_taylor_green_vortex.jl:
```diff
@@ -33,11 +33,9 @@ solver = DGMulti(polydeg = 3, element_type = Hex(), approximation_type = Polynom
-vertex_coordinates = map(x -> pi * x, vertex_coordinates) # remap the domain to [-pi, pi]^3
-mesh = VertexMappedMesh(vertex_coordinates, EToV, solver; is_periodic=(true, true, true))
-
+mesh = DGMultiMesh(solver, cells_per_dimension=(8, 8, 8),
+                     coordinates_min=(-pi, -pi, -pi), coordinates_max=(pi, pi, pi),
+                     periodicity=(true, true, true))
```
```suggestion
mesh = DGMultiMesh(solver, cells_per_dimension=(8, 8, 8),
                   coordinates_min=(-pi, -pi, -pi), coordinates_max=(pi, pi, pi),
                   periodicity=(true, true, true))
```

ranocha commented on src/solvers/dgmulti/types.jl:
```diff
@@ -76,25 +76,56 @@ function DGMulti(element_type::AbstractElemShape,
-                     is_periodic::NTuple{NDIMS, Bool} = ntuple(_->false, NDIMS)) where {NDIMS, Tv}
+  DGMultiMesh(vertex_coordinates, EToV, dg::DGMulti;
+              is_on_boundary = nothing,
+              periodic::NTuple{NDIMS, Bool} = ntuple(_->false, NDIMS)) where {NDIMS, Tv}
```
```suggestion
              periodicity::NTuple{NDIMS, Bool} = ntuple(_->false, NDIMS)) where {NDIMS, Tv}
```

ranocha commented on src/solvers/dgmulti/types.jl:
```diff
@@ -76,25 +76,56 @@ function DGMulti(element_type::AbstractElemShape,
+  DGMultiMesh(triangulateIO, dg.basis, boundary_dict)
+
+"""
+  DGMultiMesh(dg::DGMulti; cells_per_dimension,
+                coordinates_min=(-1.0, -1.0), coordinates_max=(1.0, 1.0),
+                is_on_boundary=nothing,
+                periodic=ntuple(_ -> false, NDIMS))
```
```suggestion
    DGMultiMesh(dg::DGMulti; cells_per_dimension,
                coordinates_min=(-1.0, -1.0), coordinates_max=(1.0, 1.0),
                is_on_boundary=nothing,
                periodicity=ntuple(_ -> false, NDIMS))
```

ranocha commented on src/solvers/dgmulti/types.jl:
```diff
@@ -76,25 +76,56 @@ function DGMulti(element_type::AbstractElemShape,
+Constructs a Cartesian [`DGMultiMesh`](@ref) with element type `dg.basis.elementType`. The domain is
+the tensor product of the intervals `[coordinates_min[i], coordinates_max[i]]`.
+- `is_on_boundary` specifies boundary using a `Dict{Symbol, <:Function}`
+- `periodic` is a tuple of booleans specifying periodicity = `true`/`false` in the (x,y,z) direction.
```
```suggestion
- `periodicity` is a tuple of `Bool`s specifying periodicity = `true`/`false` in the (x,y,z) direction.
```

ranocha commented on src/solvers/dgmulti/types.jl:
```diff
@@ -76,25 +76,56 @@ function DGMulti(element_type::AbstractElemShape,
+- `periodic` is a tuple of booleans specifying periodicity = `true`/`false` in the (x,y,z) direction.
+"""
+function DGMultiMesh(dg::DGMulti{NDIMS}; cells_per_dimension,
+                     coordinates_min=ntuple(_ -> -1.0, NDIMS),
+                     coordinates_max=ntuple(_ -> 1.0, NDIMS),
```
```suggestion
                     coordinates_min=ntuple(_ -> -one(real(dg)), NDIMS),
                     coordinates_max=ntuple(_ -> one(real(dg)), NDIMS),
```

ranocha commented on src/solvers/dgmulti/types.jl:
```diff
@@ -76,25 +76,56 @@ function DGMulti(element_type::AbstractElemShape,
+  vertex_coordinates, EToV = StartUpDG.uniform_mesh(dg.basis.elementType, cells_per_dimension...)
+  domain_lengths = coordinates_max .- coordinates_min
+  for i in 1:NDIMS
+    @. vertex_coordinates[i] = .5 * (vertex_coordinates[i] + 1) * domain_lengths[i] + coordinates_min[i]
```
```suggestion
    @. vertex_coordinates[i] = 0.5 * (vertex_coordinates[i] + 1) * domain_lengths[i] + coordinates_min[i]
```

jlchan commented:
@ranocha I'll add tests for the periodic SBP operators tomorrow morning and ping you when that's done.

I think I'll also clean up the type parameters since I don't appear to be using some of them...

ranocha commented:
> @ranocha I'll add tests for the periodic SBP operators tomorrow morning and ping you when that's done.
> 
> I think I'll also clean up the type parameters since I don't appear to be using some of them...

Sounds good to me, thanks :+1: You could also consider removing the type parameters from the docstrings to avoid making breaking changes when you adapt them later.

jlchan commented:
> > @ranocha I'll add tests for the periodic SBP operators tomorrow morning and ping you when that's done.
> > I think I'll also clean up the type parameters since I don't appear to be using some of them...
> 
> Sounds good to me, thanks 👍 You could also consider removing the type parameters from the docstrings to avoid making breaking changes when you adapt them later.

To me, this sounds like this process would involve
- removing type parameters from docs
- releasing a new version
- Modifying type parameters
- releasing another (technically non breaking :trollface:) version

Is that what you’re envisioning?

ranocha commented:
Basically yes. For example, I'm thinking about a docstring such as
```julia
    DGMultiMesh{NDIMS, ElemType, Nboundaries, Tv, Ti} <: AbstractMeshData{NDIMS, ElemType}
```
You could, for example, remove the implementation details `Nboundaries, Tv, Ti` from this docstring. Then, they are not part of the public API and can be changed without introducing breaking changes.

jlchan commented:
Found a bug, FDSBP doesn't seem to converge if the domain is not [-1, 1]. Working on figuring out what's going on

ranocha commented on examples/dgmulti_2d/elixir_euler_rayleigh_taylor_instability.jl:
```diff
@@ -65,15 +65,9 @@ dg = DGMulti(polydeg = 3, element_type = Quad(), approximation_type = Polynomial
-vy = map(x-> 0.5 * (1+x), vy) # map [-1, 1] to [0, 1] for single mode RTI
-vertex_coordinates = (vx, vy)
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg; is_periodic=(true,false))
+mesh = DGMultiMesh(dg, cells_per_dimension=(num_elements, 4 * num_elements),
+                     coordinates_min=(0.0, 0.0), coordinates_max=(0.25, 1.0),
+                     periodicity=(true, false))
```
```suggestion
mesh = DGMultiMesh(dg, cells_per_dimension=(num_elements, 4 * num_elements),
                   coordinates_min=(0.0, 0.0), coordinates_max=(0.25, 1.0),
                   periodicity=(true, false))
```

ranocha commented on examples/dgmulti_2d/elixir_mhd_weak_blast_wave_SBP.jl:
```diff
@@ -19,10 +19,9 @@ dg = DGMulti(polydeg=3, element_type = Quad(), approximation_type = SBP(),
-vertex_coordinates = map(x -> 2 * x, vertex_coordinates) # map domain to [-2, 2]^2
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg, is_periodic=(true, true))
-
+mesh = DGMultiMesh(dg, cells_per_dimension=cells_per_dimension,
+                     coordinates_min=(-2.0, -2.0), coordinates_max=(2.0, 2.0),
+                     periodicity=(true, true))
```
```suggestion
mesh = DGMultiMesh(dg, cells_per_dimension=cells_per_dimension,
                   coordinates_min=(-2.0, -2.0), coordinates_max=(2.0, 2.0),
                   periodicity=(true, true))
```

ranocha commented on examples/dgmulti_2d/elixir_shallowwater_source_terms.jl:
```diff
@@ -14,10 +14,10 @@ dg = DGMulti(polydeg=3, element_type = Quad(), approximation_type = SBP(),
-vertex_coordinates, EToV = StartUpDG.uniform_mesh(dg.basis.elementType, cells_per_dimension)
-vertex_coordinates = map(x -> sqrt(2) .* 0.5 .* (1 .+ x), vertex_coordinates) # map domain to [0, sqrt(2)]^2
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg, is_periodic=(true, true))
+mesh = DGMultiMesh(dg, cells_per_dimension=(8, 8),
+                     coordinates_min=(0.0, 0.0), coordinates_max=(sqrt(2), sqrt(2)),
+                     periodicity=(true, true))
```
```suggestion
mesh = DGMultiMesh(dg, cells_per_dimension=(8, 8),
                   coordinates_min=(0.0, 0.0), coordinates_max=(sqrt(2), sqrt(2)),
                   periodicity=(true, true))
```

ranocha commented:
> Found a bug, FDSBP doesn't seem to converge if the domain is not [-1, 1]. Working on figuring out what's going on

Please ping me when I shall have a look, too. (I will not be able to do it right now but maybe later, this afternoon or so.) Which setup did you use to find this problems?

ranocha commented:
I had some time to debug this and pushed a fix. Please check whether everything looks good to you, @jlchan

jlchan commented:
@ranocha thanks for catching those!

ranocha commented on src/solvers/dgmulti/sbp.jl:
```diff
@@ -298,11 +298,18 @@ function DGMultiMesh(dg::DGMultiPeriodicFDSBP{NDIMS};
   e = ones(size(rd.r))
   z = zero(e)
 
+  # VXYZ = vec.(StartUpDG.meshgrid(((coordinates_min[i], coordinates_max[i]) for i in eachindex(coordinates_min))...))
```
```suggestion
```

jlchan commented:
@ranocha I'm just cleaning up `DGMultiMesh` type parameters now and then it should be ready for another review.

ranocha commented:
Sounds good. Please ping me and @sloede when you're ready.

Do I remember correctly that you wanted to make another PR before we tag a new version of Trixi?

jlchan commented:
> Sounds good. Please ping me and @sloede when you're ready.
> 
> Do I remember correctly that you wanted to make another PR before we tag a new version of Trixi?

I shouldn't need to. I think we can release these features first then I can fix the rest of the mesh system afterwards.

ranocha commented on src/meshes/dgmulti_meshes.jl:
```diff
@@ -43,52 +39,67 @@ function Base.show(io::IO, ::MIME"text/plain", mesh::VertexMappedMesh{NDIMS, Ele
+                     periodicity=ntuple(_->false, NDIMS), kwargs...) where {NDIMS, Tv, Ti}
+
+  if haskey(kwargs, :is_periodic)
+    # TODO: DGMulti. Deprecate `is_periodic` in version 0.5
```
```suggestion
    # TODO: DGMulti, v0.5. Remove deprecated keyword
```
To unify the TODO notes related to deprecations

ranocha commented on src/meshes/dgmulti_meshes.jl:
```diff
@@ -43,52 +39,67 @@ function Base.show(io::IO, ::MIME"text/plain", mesh::VertexMappedMesh{NDIMS, Ele
+                     periodicity=(false, ), kwargs...) where {Tv, Ti}
+
+  if haskey(kwargs, :is_periodic)
+    # TODO: DGMulti. Deprecate `is_periodic` in version 0.5
```
```suggestion
    # TODO: DGMulti, v0.5. Remove deprecated keyword
```

ranocha commented on src/meshes/dgmulti_meshes.jl:
```diff
@@ -43,52 +39,67 @@ function Base.show(io::IO, ::MIME"text/plain", mesh::VertexMappedMesh{NDIMS, Ele
+  return DGMultiMesh{2, typeof(rd.elementType), typeof(md), typeof(boundary_faces)}(md, boundary_faces)
 end
 
+@deprecate VertexMappedMesh(args...; kwargs...) DGMultiMesh(args...; kwargs...)
```
```suggestion
# TODO: DGMulti, v0.5. Remove deprecated constructor
@deprecate VertexMappedMesh(args...; kwargs...) DGMultiMesh(args...; kwargs...)
```

ranocha commented on src/solvers/dgmulti/sbp.jl:
```diff
@@ -312,70 +318,79 @@ function DGMultiMesh(dg::DGMultiPeriodicFDSBP{NDIMS};
   Jf = []
 
-  is_periodic = ntuple(_ -> true, NDIMS)
+  periodic = ntuple(_ -> true, NDIMS)
```
```suggestion
  periodicity = ntuple(_ -> true, NDIMS)
```
For internal consistency

ranocha commented on src/solvers/dgmulti/sbp.jl:
```diff
@@ -312,70 +318,79 @@ function DGMultiMesh(dg::DGMultiPeriodicFDSBP{NDIMS};
   md = MeshData(VXYZ, EToV, FToF, xyz, xyzf, xyzq, wJq,
                 mapM, mapP, mapB, rstxyzJ, J, nxyzJ, Jf,
-                is_periodic)
+                periodic)
```
```suggestion
                periodicity)
```

ranocha reviewed (changes requested):
The line
https://github.com/trixi-framework/Trixi.jl/blob/e17d47bbf05695206e332b92bcc192db736d880b/docs/make.jl#L69 
should probably become something like
```
 "DGMulti mesh" => joinpath("meshes", "dgmulti_mesh.md"), 
```
I can't comment directly on the line since it's not changed here.

Otherwise, this looks good to me. Nice job, @jlchan :+1:

jlchan commented:
Odd, why didn't my commit editing `make.jl` show up? 

Thanks for adding that on your end @ranocha.

sloede commented on examples/dgmulti_1d/elixir_euler_flux_diff.jl:
```diff
@@ -11,9 +11,8 @@ equations = CompressibleEulerEquations1D(1.4)
-vertex_coordinates, EToV = StartUpDG.uniform_mesh(dg.basis.elementType, cells_per_dimension...)
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg, is_periodic=(true,))
+mesh = DGMultiMesh(dg, cells_per_dimension=(8,),
+                   coordinates_min=(-1, -1), coordinates_max=(1, 1), periodicity=(true,))
```
```suggestion
                   coordinates_min=(-1.0, -1.0), coordinates_max=(1.0, 1.0), periodicity=(true,))
```
Please also check all other modified elixirs

jlchan replied:
Done

sloede commented on examples/dgmulti_2d/elixir_euler_kelvin_helmholtz_instability.jl:
```diff
@@ -30,9 +30,7 @@ function initial_condition_kelvin_helmholtz_instability(x, t, equations::Compres
-cells_per_dimension = (32, 32)
-vertex_coordinates, EToV = StartUpDG.uniform_mesh(dg.basis.elementType, cells_per_dimension...)
-mesh = VertexMappedMesh(vertex_coordinates, EToV, dg, is_periodic=(true, true))
+mesh = DGMultiMesh(dg, cells_per_dimension=(32, 32), periodicity=(true, true))
```
IIRC, for the `TreeMesh` we support `periodicity=true` if all boundaries are to be periodic. Is it supported here as well? If yes, I'd suggest to use it

jlchan replied:
Done

sloede commented on src/meshes/dgmulti_meshes.jl:
```diff
@@ -43,52 +39,68 @@ function Base.show(io::IO, ::MIME"text/plain", mesh::VertexMappedMesh{NDIMS, Ele
 
   md = MeshData(vertex_coordinates, EToV, rd)
-  md = StartUpDG.make_periodic(md, is_periodic)
+  md = StartUpDG.make_periodic(md, periodicity)
```
E.g., if you want to enable easy setting of all periodic or non periodic, you could add a step before this call (and other ctors) similar to
```julia
if periodicity === false
  periodicity_ = ntuple(_->false, NDIMS)
elseif periodicity === true
  periodicity_ = ntuple(_->true, NDIMS)
else
  periodicity_ = periodicity
end
```
But this is just an idea and not required (feel free to resolve without further action.

sloede reviewed (changes requested):
Some minor suggestions, otherwise it LGTM

ranocha reviewed (approved):
CI failures are either spurious or caused by deprecations
